### PR TITLE
Adding Forests to IntervalRDD

### PR DESCRIPTION
**Pull request by akmorrow13**
First implementation.

### CODE REVIEW

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import org.apache.spark.SparkContext._
+import org.apache.spark.rdd.RDD
+import org.bdgenomics.utils.intervaltree.Interval
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+
+/**
+ * Companion object for building a forest from an RDD.
+ */
+object Forest extends Serializable {
```
Comment (akmorrow13) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala]: Would we like to change the name of Forest, given there is only one array now?

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  def mapValues[T2: ClassTag](f: T => T2): Forest[K, T2] = {
+    Forest(array.map(r => (r._1, f(r._2))))
+  }
+
+  /**
+   * Filters elements in array by overlapping Interval
+   * @param rr Interval to filter by
+   * @return Iterable of elements filtered by Interval rr
+   */
+  def get(rr: K): Iterable[(K, T)] = {
```
Comment (akmorrow13) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala]: @fnothaft I changed the return type from Iterable[T] to  Iterable[(K, T)] for purposes in IntervalRDD. Will this be a problem for the Tree Join? I was going to add in both but the API got a little messy.

File: utils-intervalrdd/pom.xml [hunk 3]
```diff
@@ -93,10 +93,10 @@
       <scope>test</scope>
       <type>test-jar</type>
     </dependency>
-  <dependency>
+    <dependency>
```
Comment (fnothaft) [inline utils-intervalrdd/pom.xml]: Nit: whitespace

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervalrdd/IntervalPartition.scala [hunk 4]
```diff
@@ -18,21 +18,23 @@
...
 
+import org.bdgenomics.utils.forest.Forest
 import org.bdgenomics.utils.intervaltree._
 import org.bdgenomics.utils.misc.Logging
 import scala.reflect.ClassTag
 
-protected class IntervalPartition[K <: Interval, V: ClassTag](protected val iTree: IntervalTree[K, V]) extends Serializable with Logging {
+protected class IntervalPartition[K <: Interval[K], V: ClassTag](protected val iTree: Forest[K, V]) extends Serializable with Logging {
 
+  // TODO: eliminate
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervalrdd/IntervalPartition.scala]: Can we eliminate this in this PR, or is it too much work?

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervalrdd/IntervalPartition.scala [hunk 5]
```diff
@@ -18,21 +18,23 @@
 
 package org.bdgenomics.utils.intervalrdd
 
+import org.bdgenomics.utils.forest.Forest
 import org.bdgenomics.utils.intervaltree._
 import org.bdgenomics.utils.misc.Logging
 import scala.reflect.ClassTag
 
-protected class IntervalPartition[K <: Interval, V: ClassTag](protected val iTree: IntervalTree[K, V]) extends Serializable with Logging {
+protected class IntervalPartition[K <: Interval[K], V: ClassTag](protected val iTree: Forest[K, V]) extends Serializable with Logging {
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervalrdd/IntervalPartition.scala]: This is a long line, can you break after the right paren in:

```
V: ClassTag](protected val
```

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervalrdd/IntervalRDD.scala [hunk 6]
```diff
@@ -63,10 +63,34 @@ class IntervalRDD[K <: Interval: ClassTag, V: ClassTag](
 
   /** The number of elements in the RDD. */
   override def count: Long = {
-    partitionsRDD.map(_.getTree.size).reduce(_ + _)
+    partitionsRDD.map(_.getTree.array.length).reduce(_ + _)
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervalrdd/IntervalRDD.scala]: Might be a bit cleaner to just add a `size` method on Forest, because I believe that'll be generally useful.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala (hunk 1 above)
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala]: +1, `RangeSearchableArray`?

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+
+/**
+ * Companion object for building a forest from an RDD.
+ */
+object Forest extends Serializable {
+
+  /**
+   * Sorts the RDD and collects it to build the tree.
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala]: You might want to document that this is used for the left side of the broadcast region join, since we're kind of comingling the IntervalRDD and broadcast region join code here.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala (hunk 2 above)
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala]: SGTM!

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+    optIdx.toIterable
+      .flatMap(idx => {
+        expand(rr, idx, -1) ::: expand(rr, idx + 1, 1)
+      })
+  }
+}
+
+class ForestSerializer[K <: Interval[K], T: ClassTag, TS <: Serializer[T], KS <: Serializer[K]](
+    private val kSerializer: KS,
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala]: There might be a way to get these from Kryo at runtime, which would be preferable.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Interval.scala [hunk 9]
```diff
@@ -18,12 +18,14 @@
 
 package org.bdgenomics.utils.intervaltree
 
+import scala.reflect.ClassTag
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Interval.scala]: What's this needed for here?

File: utils-intervalrdd/pom.xml (hunk 3 above)
Comment (akmorrow13) [inline utils-intervalrdd/pom.xml]: This was an indentation fix

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala (hunk 1 above)
Comment (akmorrow13) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/intervaltree/Forest.scala]: +1 for RangeSearchableArray

File: utils-intervalrdd/pom.xml (hunk 3 above)
Comment (fnothaft) [inline utils-intervalrdd/pom.xml]: Oh, sorry! You're right.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+      })
+  }
+}
+
+class RangeSearchableArraySerializer[K <: Interval[K]: ClassTag, T: ClassTag, TS <: Serializer[T], KS <: Serializer[K]](
+    private val kSerializer: KS,
+    private val tSerializer: TS) extends Serializer[RangeSearchableArray[K, T]] {
+
+  def tTag: ClassTag[T] = implicitly[ClassTag[T]]
+  def kTag: ClassTag[K] = implicitly[ClassTag[K]]
```
Comment (akmorrow13) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: @fnothaft let me know if this is what you meant for discovering types at runtime. Is there another change that should be made here?

**Comment by akmorrow13** [comment]
@fnothaft @heuermh can this make it in to the 0.2.9 release so we can use it in mango?

**Comment by heuermh** [comment]
The 0.2.9 release was already cut, so this would have to go in the next release, 0.2.10 presumably.

**Comment by jpdna** [comment]
I think I have the same question about this PR as I do here:
https://github.com/bigdatagenomics/adam/pull/1234#issuecomment-259977237
https://github.com/bigdatagenomics/adam/pull/1234#issuecomment-259988796

I didn't produce examples yet that prove these concerns are real  - for now I just want to see if you all think these points need further investigation.

**Comment by akmorrow13** [comment]
@jpdna isnt this solved if you don't set orientation? I imagine that if you ran a query on a ReferenceRegion whose orientation is set, by definition you wouldn't want reads of the other orientation. Is this incorrect?

**Comment by fnothaft** [comment]
> @jpdna isnt this solved if you don't set orientation? I imagine that if you ran a query on a ReferenceRegion whose orientation is set, by definition you wouldn't want reads of the other orientation. Is this incorrect?

I agree with @akmorrow13 on this point. 

I think that @jpdna may have caught a corner case in the second comment, but I need to take more of a look/write unit tests.

**Comment by akmorrow13** [comment]
@fnothaft @jpdna  I'll add a test for bigdatagenomics/adam#1234 (comment) . Unfortunately these cases were where the Interval tree came in.

**Comment by fnothaft** [comment]
> @fnothaft @jpdna I'll add a test for bigdatagenomics/adam#1234 (comment) . Unfortunately these cases were where the Interval tree came in.

Yeah I'm pretty sure that the logic in the current code is correct, but without writing a test or a proof, YMMV. I worked out a few cases mentally and they seemed right, but TBH I'm zonked enough that I would not be surprised to see a bug in the code. Thanks for taking that on @akmorrow13!

**Comment by jpdna** [comment]
> by definition you wouldn't want reads of the other orientation

I'm not sure this is true, and to be clear here we are talking here specifically about joining reference regions based on interval overlap not about if next gen reads are in reverse or forward.  Say you are joining a VCF to a exon set:  the VCF has all alleles represented on forward strand by definition.  The exon feature set will likely have have strand set, in fact you need the strand from the transcripts to know when you need to complement the variants ( I think ).  If we don't set strand orientation when loading exon data then probably we should.  @heuermh  can you take a look a this too to see if I am getting strand confused?

**Comment by fnothaft** [comment]
@jpdna Alyssa can correct me if I'm misreading her comment, but I read "by definition" as referring to "by the definition of the ReferenceRegion class", not as "by definition, for all genomic data". IIRC, all of our apply methods on the Avro models make INDEPENDENT stranded regions.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala [hunk 11]
```diff
@@ -16,14 +16,14 @@
...
-package org.bdgenomics.utils.intervaltree
+package org.bdgenomics.utils.rangearray
 
 /**
  * An interval is a 2-dimensional 0-based coordinate consisting of a closed start value and open end value.
  * Each 2-dimensional coordinate has a defined width.
  * This can be used to express a region of a genome, a transcript, a gene, etc.
  */
-trait Interval {
+trait Interval[T <: Interval[T]] extends Comparable[T] {
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala]: I don't think this type needs to be recursive. I think the `overlaps` and `compareTo` functions should take params of type `Interval` and not `T`. However, is this necessary to support `Interval`s that exist on different coordinate planes? I.e., going from supporting a single contig to multiple contigs? If so, I can get behind the recursive type.

Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/Interval.scala]: Actually yeah, I just checked looked at some of the code below and I'm pretty sure that's why. If so, savvy! I like it, let's just document it in the scaladoc for the trait for posterity.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+ * Alas, we have no trees anymore.
+ * I blame global warming.
+ *
+ * @param arr An array of values for the left side of the join. We require
+ *   this array to be sorted.
+ */
+case class RangeSearchableArray[K <: Interval[K], T: ClassTag](arr: Array[(K, T)], sorted: Boolean = false) {
+
+  // ensure that array is sorted
+  val array =
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: This looks new. Is this for the append case? If so, let's make a few changes:

* Make `val array` private
* Make `RangeSearchableArray`:
  * A class and not a case class
  * extend Serializable

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+      idx >= length ||
+      !rr.overlaps(array(idx)._1)) {
+      list
+    } else {
+      expand(rr, idx + step, step, array(idx) :: list)
+    }
+  }
+
+  /**
+   * Insert Iterator of (K,V) items into existing RangeSearchableArray
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Nit: insert `an`  in `Insert Iterator`, period at end of line, space after line.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+   * @param kvs (K,V) tuples to insert into RangeSearchableArray
+   * @return new RangeSearchableArray with inserted values
+   */
+  def insert(kvs: Iterator[(K, T)], sorted: Boolean = false): RangeSearchableArray[K, T] = {
+
+    // sort kvs if not yet sorted
+    val sortedKvs =
+      if (sorted) kvs.toArray
+      else kvs.toArray.sortBy(_._1)
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Can you open a ticket to look into efficient ways to do the sort-on-insert? This looks like a good start but I have some more ideas.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+
+    // initialize empty array to merge two arrays into
+    val allSorted: Array[(K, T)] = new Array[(K, T)](array.length + sortedKvs.length)
+    var i = 0
+    var j = 0
+
+    // merge two sorted arrays
+    Array.range(0, allSorted.length).foreach(k => {
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: `Array.range(0, allSorted.length)` -> `allSorted.indices`.

Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Also, this looks good for now, but we can probably eke out performance gains by writing this as an `@tailrec` function due to the way `@tailrec`/`while` loops are implemented in Scala vs `foreach`es. Not important for a first cut, but let's come back to that.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+          j = j + 1
+          sortedKvs(j - 1)
+        })
+    })
+
+    RangeSearchableArray(allSorted)
+  }
+
+  /**
+   * Filters items in RangeSearchableArray based on predicate on (K,V) tuples
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Nit: period at EOL, newline after.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+   * @param pred predicate to filter elements by
+   * @return new RangeSearchableArray with filtered elements
+   */
+  def filter(pred: ((K, T)) => Boolean): RangeSearchableArray[K, T] = {
+    RangeSearchableArray(array.filter(r => pred(r._1, r._2)))
+  }
+
+  /**
+   * Maps values from T to T2
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Nit: period at EOL (should be a sentence), newline after.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+   * @param f Function mapping T to T2
+   * @tparam T2 new type to map values to
+   * @return new RangeSearchableArray with mapped values
+   */
+  def mapValues[T2: ClassTag](f: T => T2): RangeSearchableArray[K, T2] = {
+    RangeSearchableArray(array.map(r => (r._1, f(r._2))))
+  }
+
+  /**
+   * Filters elements in array by overlapping Interval
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Nit: period at EOL (should be a sentence), newline after.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+   * @param rr Interval to filter by
+   * @return Iterable of elements filtered by Interval rr
+   */
+  def get(rr: K): Iterable[(K, T)] = {
+
+    val optIdx = binarySearch(rr)
+
+    optIdx.toIterable
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Do `toIterable` last.

File: utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+    }
+  }
+
+  @tailrec private def expand(rr: K,
+                              idx: Int,
+                              step: Int,
+                              list: List[(K, T)] = List.empty): List[(K, T)] = {
+    if (idx < 0 ||
+      idx >= length ||
+      !rr.overlaps(array(idx)._1)) {
```
Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Dropping a note to myself here. Related to @jpdna's question about whether this code can be proven to work (with the case he gave) and the second question about data mapped to a specific strand, I've thought about this and:

1. This code can be proven to work!
2. ...unless you have switches in strandedness...
3. But! There is a simple fix for said issue.

I'll open a PR to address this. What I'll add is:

1. a function in Interval called `covers`, which by default calls to `overlaps`. This is a less strict version of `overlaps` that ignores secondary predicates in the coordinate space (like our strandedness thing).
2. the proof

Comment (akmorrow13) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: @fnothaft I added a test for the concern of @jpdna in RangeSearchableArraySuite

Comment (jpdna) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Can you point me to the line of code with this new test @akmorrow13   ?

Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: On phone, so hard to search/get a link, but if you look for a variable named something like `longRegion` in `RangeSearchableArraySuite`, that should be the one.

Comment (jpdna) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: If I make very slightly change the test, for which assertion is at line 128 in RangeSearchableArraySuite.scala to:
```
  sparkTest("verify RangeSearchableArray fetches all valid ranges") {
    val longRegion = Region(1L, 200L)
    val rdd = sc.parallelize(Seq((longRegion, 0),
      (Region(2L, 10L), 1),
      (Region(18L, 24L), 2),
      (Region(28L, 32L), 3),
      (Region(32L, 40L), 4),
      (Region(40L, 60L), 5)))

```
to cause the "long" region to start at (now 1L) before the next region that (now starts at 2L and ends at 10L), the  the test fails.   Unless I am missing something, I don't see why this change to the test should have changed the resulting overlap count away from the current 3 for the test query.

Comment (fnothaft) [inline utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala]: Yeah, there's a bit more to the fix. I've got a patch locally (with the proof) that resolves this. I'm just putting the finishing on the proof, will push shortly.

**Comment by fnothaft** [comment]
(Somewhat informal) Proof at https://github.com/akmorrow13/utils/pull/1.

**Comment by heuermh** [comment]
@akmorrow13 looks like you may have lost my doc suggestions from https://github.com/akmorrow13/utils/pull/1

**Comment by akmorrow13** [comment]
@heuermh I merged the PR

**Comment by heuermh** [comment]
Right, but without my doc suggestions.  :)  Should I re-do them here?

**Comment by akmorrow13** [comment]
Shoot, I'm not sure. I tried reverting the PR to get them back but it looks like everything is gone now.

**Comment by fnothaft** [comment]
> Shoot, I'm not sure. I tried reverting the PR to get them back but it looks like everything is gone now.

I'll open a PR with @heuermh's doc suggestions.

**Comment by fnothaft** [comment]
Addressed @heuermh's documentation comments in https://github.com/akmorrow13/utils/pull/3.

**Comment by fnothaft** [comment]
This LGTM, but I am going to test it with #1234 before merging. @jpdna and @heuermh can you review/approve this PR?

**Comment by heuermh** [comment]
Since this is utils, would it be ok to add the new methods to a `Range` trait for `RangePartition`, `RangeSearchableArray`, and bigdatagenomics/adam#1234, and keep `Interval` and related around?  With some additional polish and performance tuning, perhaps `IntervalTree` might still be useful for other use cases.

**Comment by fnothaft** [comment]
@heuermh not entirely sure I understood you, is what you're saying:

> Since this is utils, would it be ok to add the new methods to a `Range` trait for `RangePartition`, `RangeSearchableArray`, and bigdatagenomics/adam#1234, and keep `Interval` and related around?

Suggesting to:

* move the `covers` and `distance` traits to a `Range` trait (that extends `Interval`)
* and keep `Interval` and `IntervalTree` around?

From my side, moving the new methods to a `Range` trait and keeping `Interval` unchanged sound a-OK. I'd rather not keep the `IntervalTree` around, since it'd be effectively unused code, once we chop `IntervalTree` out in Mango and bigdatagenomics/adam#1234 merges.

That being said, what might be a reasonable compromise would be to rewrite history so that we rm `IntervalTree` in one commit, and then do all the other changes. This way, if we wanted to bring back and tune `IntervalTree` at a later date, we'd just have to revert a single, simple commit. After having taken a long look at this code path, I guess I'm not terribly bullish on interval trees as a data structure, for a couple of reasons:

* The worst-case lookup cost in an interval tree is the same as the worst-case lookup case in said array (both are `O(log n)`).
* The memory footprint of this data structure is much better, especially from a GC perspective. Even better, we should be able to move this code to use the `unsafe` library for the core array, which should give both a performance boost (we bound check the array indices before all accesses already, so moving to `unsafe` would eliminate the repeated bound checks) and would allow us to move a big array off-heap (which has huge GC implications, esp. for algos that make use of large broadcasts, such as BQSR).
* Not quite the same as the above, but this structure is also much cheaper to serialize since we don't need to either:
  * Write a complex serializer that serializes the tree structure, or
  * Flatten the tree when serializing and then rebuild the tree when deserializing
* We can also do a nifty optimization where we cache prior binary search results on key lookup, which should improve `get` performance.

The interval tree is _definitely_ better for single key insert. That being said, I _think_ that with a minor optimization, the bulk insert costs are roughly the same. Specifically, if a bulk insert with keys strictly before the start of the array or after the end of the array, we just have to copy the bulk inserted keys to the start/end. This obviously doesn't cover all cases, but I think it covers the important case from Mango.

Thoughts?

**Comment by heuermh** [comment]
Thanks for the detailed response.

Alternatively, with new methods on `Interval`, what would you think of a rename `RangePartition` and `RangeSearchableArray` &rarr; `IntervalPartition` and `IntervalArray`?

In any case, rewriting history to rm `IntervalTree` in one commit is a good idea.

**Comment by fnothaft** [comment]
> Alternatively, with new methods on Interval, what would you think of a rename RangePartition and RangeSearchableArray → IntervalPartition and IntervalArray?

SGTM! I just threw out the `RangeSearchableArray` name.

I can take these changes on.

**Comment by fnothaft** [comment]
> In any case, rewriting history to rm IntervalTree in one commit is a good idea.

Ah, just tried doing this and alas, it isn't possible since `IntervalTree` ties in to `IntervalPartition` and `IntervalRDD`.

**Comment by fnothaft** [comment]
Also, I figured out how to rewrite history. Will PR in the AM...

**Comment by akmorrow13** [comment]
Sorry I dropped out of the convo for a day. Are we ready for me to rebase so we can get this merged?

**Comment by fnothaft** [comment]
Don't worry about rebasing it. I have a branch locally with almost all of @heuermh's comments RE naming incorporated. I've got about 15min more work to do on it and will push it tomorrow.

**Comment by akmorrow13** [comment]
sorry to drop the bomb on that. Thanks @fnothaft for that!

**Comment by fnothaft** [comment]
No problem @akmorrow13! I hope you had a good Thanksgiving. Closing this, and continuing in #94.
